lib/SILOptimizer/Mandatory/TFPartition.cpp

```diff
@@ -1002,21 +1006,39 @@ void TFFunctionPartition::markBlock(SILBasicBlock *BB) {
       // aren't control-dependent on it.  It may be control-dependent on
       // something though.
       if (numTensorSuccs == 1) {
+        assert(tensorCodeBlocks.postDominates(thisBB, pred));
+        assert(tensorCodeBlocks.postDominates(BB, pred));
         worklist.push_back(pred);
         continue;
       }
 
-      // Otherwise, check to see if BB is post-dominated by thisBB, but not by
-      // pred - in other words that it is the post dominance frontier for BB.
-      // If so, that means that the terminator in pred controls whether BB is
-      // executed, so it must be marked.
+      // Otherwise, we already know by construction that BB post-dominates
+      // thisBB -- now we check to see if BB also post-dominates pred. If not,
+      // pred is the post dominance frontier for BB, and that means that the
+      // terminator in pred controls whether BB is executed, so it must be
+      // marked.  For example, BB is control-dependent on pred in this CFG:
+      //
+      //    pred
+      //    /  \
+      //   ..  thisBB
+      //    \   |
+      //    ..  BB
+      //      \ |
+      //       ..
       if (!tensorCodeBlocks.postDominates(BB, pred)) {
         auto predTerm = pred->getTerminator();
         if (isa<BranchInst>(predTerm) || isa<CondBranchInst>(predTerm)) {
           markInstruction(*predTerm, Marking::Copy);
           continue;
         }
 
+        if (isa<SwitchEnumInst>(predTerm)) {
```

Comment:
Is there anything specific about SwitchEnum for this, or can it be generalized to support any non-branch terminator?

Reply:
I haven't looked at the other term insts to conclude that they can all be handled in this way. I'll file a bug on this subject. Util we finish that, I feel it'd be good to crash with an "unsupported" message than to risk getting into unknown unknowns (e.g. generate bad code).

Reply:
Filed https://bugs.swift.org/browse/SR-8167